### DISCUSSION TITLE
Performance implications of Python code

### DISCUSSION THREAD

**Discussion by snnsnn** [geldata/gel#4449] [Q&A] [answered]
I really liked EdgeDB's ambitions and design decisions and planning on using it but not sure about the performance of Python code since so many parts are written in Python. Why Python? You have parser written in Rust, any plan on using Rust for other parts? Why not go full Rust?

**Comment by elprans** [comment, marked as answer]
> not sure about the performance of Python code

There is very little Python code in the critical path of actually running queries (see https://github.com/edgedb/imdbench for a realistic benchmark).  The "slow" part of EdgeDB is only engaged when the server sees a particular query for the first time.

> Why Python?

Mostly historical reasons (EdgeDB has been in development before Rust became mature), but also see @msullivan's [answer in an earlier discussion](https://github.com/edgedb/edgedb/discussions/3442#discussioncomment-2155778).

> any plan on using Rust for other parts?

We will most likely replace the non-compiler parts with Rust, though mostly for safety and correctness reasons rather than speed, because those parts are already fast enough (on the order of ~100us of overhead per query).

> Why not go full Rust?

That would be counter-productive as we would have to essentially stop all meaningful language and schema development for a rewrite that could easily take several developer-years.  Rust also has drawbacks and it's not clear that it would be a more productive language to write a complex compiler in.
